File: docs/palindromes-in-peg/scaffold_round_lazy.py
"""Demand-driven construction of the observable part of a packed round.

The same virtual-node equations as scaffold_round are resolved only when a
query needs them. Persisted fields are closed under all old-node queries
reachable from acceptance. No input length or execution trace is specialized.
"""
from scaffold_round import RoundBuilder, Address
from scaffold_circuit import Value, choose_pointer
from symbolic_sca2peg import Scaffold, TRUE, FALSE, NULL, SELF, pointer


class MissingField(Exception):
  def __init__(self, kind, slot, key):
    self.field = kind, slot, key


class SlotFields(dict):
  def __init__(self, kind, slot):
    self.kind, self.slot = kind, slot

  def __missing__(self, key):
    raise MissingField(self.kind, self.slot, key)


class DemandRoundBuilder(RoundBuilder):
  def __init__(self, stages, input_symbols=None, alphabet=None):
    super().__init__(stages, input_symbols, alphabet)
    self.slot_labels = [SlotFields("label", i) for i in self.tags]
    self.slot_pointers = [SlotFields("pointer", i) for i in self.tags]
    self.memo = {}
    self.pending, self.requested = [], set()

  def need(self, kind, slot, key):
    field = kind, slot, key
    if field not in self.requested:
      self.requested.add(field)
      self.pending.append(field)

  def read(self, target, key):
    if target.prior != NULL:
      for slot in self.possible_tags(target.tag):
        if target.tag.eq(slot) != FALSE: self.need("label", slot, key)
    return super().read(target, key)

  def follow(self, target, key):
    if target.prior != NULL:
      for slot in self.possible_tags(target.tag):
        if target.tag.eq(slot) != FALSE: self.need("pointer", slot, key)
    return super().follow(target, key)

  def resolve(self, field):
    work, active = [field], {field}
    while work:
      kind, slot, key = current = work[-1]
      fields = self.slot_labels if kind == "label" else self.slot_pointers
      if key in fields[slot]:
        active.remove(work.pop())
        continue
      source = self.stages[slot]
      expression = (source.labels if kind == "label" else source.pointers)[key]
      root = Address({}, pointer(()), Value.constant(self.tags[-1]).recode(self.tags)) if slot == 0 else \
             Address({slot - 1: TRUE}, NULL, self.zero_tag)
      try:
        fields[slot][key] = self.translate(expression, slot, root, self.memo.setdefault(slot, {}))
        active.remove(work.pop())
      except MissingField as missing:
        if missing.field in active:
          raise ValueError("cyclic construction query in packed round") from missing
        active.add(missing.field)
        work.append(missing.field)
    return fields[slot][key]

  def build(self):
    accepting = self.stages[-1].accepting
    self.need("label", self.tags[-1], accepting)
    while self.pending:
      kind, slot, key = field = self.pending.pop()
      value = self.resolve(field)
      if kind == "label":
        name = self.label(slot, key)
        self.initial[name], self.labels[name] = self.stages[slot].initial[key], value
      else:
        target, tag = value.prior, value.tag
        for local, guard in value.local.items():
          target = choose_pointer(guard, SELF, target)
          tag = Value.select(guard, Value.constant(local), tag)
        self.pointers[self.field(slot, key)] = target
        for bit, expression in enumerate(tag.recode(self.tags).bits):
          name = self.tag_bit(slot, key, bit)
          self.initial[name], self.labels[name] = False, expression
    return Scaffold(self.initial, self.labels, self.pointers,
                     self.label(self.tags[-1], accepting), self.alphabet)


def pack_round_lazy(stages, input_symbols=None, alphabet=None):
  return DemandRoundBuilder(stages, input_symbols, alphabet).build()

File: scala/pal/src/main/scala/pal/ScaffoldRoundLazy.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable

import Expr.{TRUE, FALSE, NULL}

/** Demand-driven construction of the observable part of a packed round.
  *
  * The same virtual-node equations as scaffold_round are resolved only when a
  * query needs them. Persisted fields are closed under all old-node queries
  * reachable from acceptance. No input length or execution trace is specialized.
  *
  * Port of `scaffold_round_lazy.py`.
  */
object ScaffoldRoundLazy {

  def packRoundLazy(
      stages: Seq[Scaffold],
      inputSymbols: Option[Seq[collection.Map[Char, Expr]]] = None,
      alphabet: Option[String] = None
  ): Scaffold = {
    new DemandRoundBuilder(stages, inputSymbols, alphabet).build()
  }
}

/** Whether a slot field is a label or a pointer (Python `"label"` / `"pointer"`). */
enum FieldKind {
  case Label, Pointer
}

/** A slot field that a translation depends on. */
final case class SlotField(kind: FieldKind, slot: Int, key: String)

/** Raised by a slot lookup whose field has not been resolved yet (Python `MissingField`).
  *
  * A control-flow signal: it carries no stack trace.
  */
final class MissingField(val field: SlotField) extends RuntimeException(s"missing ${field.kind} field ${field.slot}.${field.key}") {
  override def fillInStackTrace(): Throwable = this
}

/** A `RoundBuilder` whose slot tables fill on demand, driven by acceptance. */
final class DemandRoundBuilder(
    stageSeq: Seq[Scaffold],
    inputSymbolSeq: Option[Seq[collection.Map[Char, Expr]]] = None,
    alphabetOverride: Option[String] = None
) extends RoundBuilder(stageSeq, inputSymbolSeq, alphabetOverride) {
  import FieldKind.{Label, Pointer}
  import RoundBuilder.label

  for (_ <- tags) {
    slotLabels += mutable.LinkedHashMap.empty
    slotPointers += mutable.LinkedHashMap.empty
  }

  /** One translation memo per slot (Python `self.memo.setdefault(slot, {})`). */
  private val memos = mutable.HashMap.empty[Int, java.util.IdentityHashMap[Expr, RoundValue]]
  private val pending = mutable.ArrayBuffer.empty[SlotField]
  private val requested = mutable.HashSet.empty[SlotField]

  override protected def slotLabel(slot: Int, key: String): Expr = {
    slotLabels(slot).getOrElse(key, throw new MissingField(SlotField(Label, slot, key)))
  }

  override protected def slotPointer(slot: Int, key: String): Address = {
    slotPointers(slot).getOrElse(key, throw new MissingField(SlotField(Pointer, slot, key)))
  }

  def need(kind: FieldKind, slot: Int, key: String): Unit = {
    val field = SlotField(kind, slot, key)
    if (requested.add(field)) { pending += field }
  }

  override def read(target: Address, key: String): Expr = {
    if (target.prior != NULL) {
      for (slot <- possibleTags(target.tag)) {
        if (target.tag.eqTo(slot) != FALSE) { need(Label, slot, key) }
      }
    }
    super.read(target, key)
  }

  override def follow(target: Address, key: String): Address = {
    if (target.prior != NULL) {
      for (slot <- possibleTags(target.tag)) {
        if (target.tag.eqTo(slot) != FALSE) { need(Pointer, slot, key) }
      }
    }
    super.follow(target, key)
  }

  private def resolved(field: SlotField): Boolean = {
    field.kind match {
      case Label => slotLabels(field.slot).contains(field.key)
      case Pointer => slotPointers(field.slot).contains(field.key)
    }
  }

  private def lookup(field: SlotField): RoundValue = {
    field.kind match {
      case Label => slotLabels(field.slot)(field.key)
      case Pointer => slotPointers(field.slot)(field.key)
    }
  }

  private def rootFor(slot: Int): Address = {
    if (slot == 0) { previousRoot() } else { Address(VectorMap(slot - 1 -> TRUE), NULL, zeroTag) }
  }

  /** Translate `field`, first resolving every unresolved field its translation queries. */
  def resolve(field: SlotField): RoundValue = {
    val work = mutable.ArrayBuffer(field)
    val active = mutable.HashSet(field)
    while (work.nonEmpty) {
      val current = work.last
      if (resolved(current)) {
        active -= work.remove(work.length - 1)
      } else {
        val stage = stages(current.slot)
        val memo = memos.getOrElseUpdate(current.slot, new java.util.IdentityHashMap[Expr, RoundValue]())
        try {
          current.kind match {
            case Label =>
              val value = translate(stage.labels(current.key), current.slot, rootFor(current.slot), memo)
              slotLabels(current.slot)(current.key) = value.asInstanceOf[Expr]
            case Pointer =>
              val value = translate(stage.pointers(current.key), current.slot, rootFor(current.slot), memo)
              slotPointers(current.slot)(current.key) = value.asInstanceOf[Address]
          }
          active -= work.remove(work.length - 1)
        } catch {
          case missing: MissingField =>
            if (active.contains(missing.field)) {
              throw new IllegalArgumentException("cyclic construction query in packed round", missing)
            }
            active += missing.field
            work += missing.field
        }
      }
    }
    lookup(field)
  }

  override def build(): Scaffold = {
    val accepting = stages.last.accepting
    need(Label, tags.last, accepting)
    while (pending.nonEmpty) {
      val field = pending.remove(pending.length - 1)
      val value = resolve(field)
      field.kind match {
        case Label => emitLabel(field.slot, field.key, value.asInstanceOf[Expr])
        case Pointer => emitPointer(field.slot, field.key, value.asInstanceOf[Address])
      }
    }
    new Scaffold(initial, labels, pointers, label(tags.last, accepting), alphabet)
  }
}
